server/src/pages/shot/text-tool.js

```diff
@@ -0,0 +1,236 @@
+const React = require("react");
+const PropTypes = require("prop-types");
+const { Localized } = require("fluent-react/compat");
+const sendEvent = require("../../browser-send-event.js");
+const { Selection } = require("../../../shared/selection");
+
+// Initial position of displayed text input area
+let INIT_LEFT;
+let INIT_TOP;
+const INIT_WIDTH = 100;
+
+// Text input height is determinedd by text size
+// Additional padding added to text size to compute text input height
+const TEXT_INPUT_PADDING = 3;
+const FONT_STYLE = "sans-serif";
+const FONT_SIZE = 36;
+const FONT_WEIGHT = 900;
+const BKG_COLOR = "rgba(0, 0, 0, 0.6)";
+const FONT_COLOR = "#FF0";
+
+let isMousedown;
+let isFirstMouseDown;
+let mousedownPosition;
+let mousedownSelection;
+
+exports.TextTool = class TextTool extends React.Component {
+  constructor(props) {
+    super(props);
+    this.canvas = React.createRef();
+    this.el = React.createRef();
+    this.state = {
+      textSelection: null,
+      textContent: ""
+    };
+    this.canvasWidth = parseInt(props.baseCanvas.style.width, 10);
+    this.canvasHeight = parseInt(props.baseCanvas.style.height, 10);
+  }
+
+  componentDidMount() {
```

Comment:
Because `textSelection` is null at this point, the initial `render()` did not render a canvas.  That's why `this.canvas.current` is null here.

Reply:
Thanks @chenba - makes sense the life cycle sequence is render -> ComponentDidMount. Moving the initialization of textSelection to ComponentWillMount resolved the issue of this.canvas.current null error

Reply:
@punamdahiya `componentWillMount` will be removed, and shouldn't be used; https://reactjs.org/docs/react-component.html#unsafe_componentwillmount

Reply:
@chenba Thanks for letting know, as recommended will initialize textSelection state in constructor itself.